### ISSUE TITLE
Prakriya doesn't show all rules involved in the process?

### ISSUE THREAD

**Issue by vipranarayan14** [ambuda-org/vidyut#44]
In the derivation of बुध्यात् (बुधँ अवगमने 01.0994):

- [पुगन्तलघूपधस्य च](https://ashtadhyayi.com/sutraani/7/3/86) has prApti but is not shown.
- [क्ङिति च](https://ashtadhyayi.com/sutraani/1/1/5) prohibits पुगन्तलघूपधस्य च but is not shown.

Without the presence of atleast the क्ङिति च, it is difficult to understand why there is no guNa in the form.

![image](https://user-images.githubusercontent.com/17687018/217571371-6995b3ed-3628-4ffc-993b-b8f062f0a4fc.png)

Dr Dhaval's Prakriyāpradarśinī shows  क्ङिति च:

![image](https://user-images.githubusercontent.com/17687018/217570674-7d51912f-131e-4b1f-9b42-b971870f8653.png)

[If this is intended](https://github.com/ambuda-org/vidyut/issues/15#issuecomment-1367094617), I understand.

**Comment by akprasad** [comment]
Thanks for filing this! I think we should add क्ङिति च here. The current logic, in pseudocode, is something like:

```
if is_knit {
  return
}
if sarvadhatuka_ardhadhatukayoh {
  do_guna()
  mark("7.3.84")
} else if puganta_laghupadhasya {
  do_guna()
  mark("7.3.86")
}
```

It should instead be:

```
if sarvadhatuka_ardhadhatukayoh {
  if is_knit {
   mark("1.1.5")
  } else {
    do_guna()
    mark("7.3.84")
  }
} else if puganta_laghupadhasya {
  if is_knit {
   mark("1.1.5")
  } else {
    do_guna()
    mark("7.3.86")
  }
}
```

Do you want to try making the code change?

**Comment by vipranarayan14** [comment]
Yes! It will allow me to understand the inner workings of vidyut.

**Comment by vipranarayan14** [comment]
> The current logic, in pseudocode, is something like:
> 
> ```
> if is_knit {
>   return
> }
> if sarvadhatuka_ardhadhatukayoh {
>   do_guna()
>   mark("7.3.84")
> } else if puganta_laghupadhasya {
>   do_guna()
>   mark("7.3.86")
> }
> ```

The [actual code used for applying guna](https://github.com/vipranarayan14/vidyut/blob/ede1e45f89af8e164eec6e2f624fa5c5f9485f0c/vidyut-prakriya/src/angasya/guna_vrddhi.rs#L179) is different from what you have described here. It seems it is more like:

```
if laghu_upadha && can_use_guna {
    do_guna()
    mark("7.3.86")
} else if ik_anta && can_use_guna {
    do_guna()
    mark("7.3.84")
}
```

**Comment by akprasad** [comment]
Thanks for checking -- I was speaking loosely to keep the description simple, but I'll be more precise in the future.

In this case, the control flow is quite complex, but I think we can try pushing it in:

```
if laghu_upadha {
  if can_use_guna {
    mark("7.3.86")
  } else {
    mark("1.1.5");
  } 
}
```

Feel free to heavily modify this function, too -- I dislike the control flow here, and if you can find a cleaner approach, I'm all for it.

**Comment by vipranarayan14** [comment]
> Thanks for checking -- I was speaking loosely to keep the description simple, but I'll be more precise in the future.
> 
> In this case, the control flow is quite complex, but I think we can try pushing it in:
> 
> ```
> if laghu_upadha {
>   if can_use_guna {
>     mark("7.3.86")
>   } else {
>     mark("1.1.5");
>   } 
> }
> ```

I did the same thing. But I couldn't make it pass all the tests. After that, I couldn't find time to work on it. I'll try it again sometime tomorrow and let you know.

> Feel free to heavily modify this function, too -- I dislike the control flow here, and if you can find a cleaner approach, I'm all for it.

Thanks for the permission 🙂. Some people don't like others disturbing their codebase. However, I'll modify them in small steps so that I don't break anything accidentally. (Your extensive suite of tests wouldn't allow me to break anything anyway.)
